### PR TITLE
feat(auth): add grpctransport package

### PR DESCRIPTION
**Pull request by codyoss**
This package is the analog to https://pkg.go.dev/google.golang.org/api/transport/grpc.

### CODE REVIEW

File: auth/grpctransport/testdata/echo.proto
Comment (noahdietz) [inline auth/grpctransport/testdata/echo.proto:1]: How do we plan to keep this file auto generated with the latest? Can you include some steps in a README to regenerate them? Should we just automate regeneration via github actions somehow?

Thanks.

**Review by noahdietz — COMMENTED** [review]
LGTM just wondering how we keep the test bindings up to date in the event new proto/grpc features land and we need to regen.

File: auth/grpctransport/directpath.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	if !checkDirectPathEndPoint(endpoint) {
+		return false
+	}
+	if b, _ := strconv.ParseBool(os.Getenv(disableDirectPathEnvVar)); b {
+		return false
+	}
+	return true
+}
+
+func checkDirectPathEndPoint(endpoint string) bool {
```
Comment (quartzmo) [inline auth/grpctransport/directpath.go:43]: I think maybe this is the only func in this file that is tested? Any plans to add tests for the others?

Comment (quartzmo) [reply auth/grpctransport/directpath.go:43]: Nm, I think `TestDialWithDirectPathEnabled` covers the rest?

File: auth/grpctransport/grpctransport.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+	}
+	if opts.PoolSize == 0 || opts.PoolSize == 1 {
+		conn, err := dial(ctx, secure, opts)
+		if err != nil {
+			return nil, err
+		}
+		return &singleConnPool{conn}, nil
+	}
+	pool := &roundRobinConnPool{}
+	for i := 0; i < opts.PoolSize; i++ {
```
Comment (quartzmo) [inline auth/grpctransport/grpctransport.go:164]: What do we get if PoolSize is < 0? An empty pool? Should PoolSize be validated?

File: auth/grpctransport/grpctransport.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+		tOpts.DefaultEndpoint = io.DefaultEndpoint
+		tOpts.DefaultMTLSEndpoint = io.DefaultMTLSEndpoint
+	}
+	transportCreds, endpoint, err := transport.GetGRPCTransportCredsAndEndpoint(tOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	if !secure {
+		transportCreds = grpcinsecure.NewCredentials()
```
Comment (quartzmo) [inline auth/grpctransport/grpctransport.go:192]: Would it make sense to do this before the call to `GetGRPCTransportCredsAndEndpoint`?

File: auth/grpctransport/grpctransport.go (hunk 2 above)
Comment (codyoss) [reply auth/grpctransport/grpctransport.go:164]: We default to 1, see code 8 lines up.

Comment (quartzmo) [reply auth/grpctransport/grpctransport.go:164]: You mean this line? 

```
if opts.PoolSize == 0 || opts.PoolSize == 1 {
```

What if `opts.PoolSize` is `-1` for example? It's a signed int, isn't it?

File: auth/grpctransport/grpctransport.go (hunk 3 above)
Comment (quartzmo) [reply auth/grpctransport/grpctransport.go:192]: Great, thanks.

File: auth/grpctransport/grpctransport.go (hunk 2 above)
Comment (codyoss) [reply auth/grpctransport/grpctransport.go:164]: ah, I misread your comment you are correct. Do you think we should validate it or just default to 1 in these cases? Either seems fine to me. I think the original impl would have returned an empty pool that never works, but I don't think that is desired.

Comment (quartzmo) [reply auth/grpctransport/grpctransport.go:164]: I would just default to 1, and change the docs to say: "If unset or less than 1, the default is 1."

Comment (codyoss) [reply auth/grpctransport/grpctransport.go:164]: done

**Review by quartzmo — APPROVED** [review]

**Pull request merged**